tiledb/sm/buffer/buffer.h

```diff
@@ -418,6 +435,226 @@ class ConstBuffer : public BufferBase {
+   * function.
+   *
+   * @param size The size of the buffer.
+   * @param alloc Optional allocator for the buffer.
+   */
+  SerializationBuffer(size_t size, const allocator_type& alloc = {})
+      : buffer_owner_(size, alloc)
+      , buffer_(buffer_owner_) {
+  }
+
+  /**
+   * Shorthand constructor for a non-owned buffer to a given pointer-size pair.
+   *
+   * @param data The data of the buffer.
+   * @param size The size of the buffer.
+   */
+  SerializationBuffer(
+      NonOwnedMarker,
+      const void* data,
+      size_t size,
+      const allocator_type& alloc = {})
+      : buffer_owner_(alloc)
+      , buffer_(static_cast<const char*>(data), size) {
+  }
+
+  /**
+   * Returns the buffer's allocator. This is required to make the type
+   * allocator-aware.
+   */
+  inline allocator_type get_allocator() const noexcept {
+    return buffer_owner_.get_allocator();
+  }
+
+  bool is_owned() const {
+    return buffer_.data() == buffer_owner_.data();
+  }
+
+  /**
+   * Assigns a new owned buffer to the object.
+   *
+   * @param iter The iterator range to copy to the serialization buffer.
+   */
+  template <class Iter>
+  inline void assign(const Iter& iter) {
+    // Clear vector and deallocate its buffer.
+    buffer_owner_.clear();
+    buffer_owner_.shrink_to_fit();
+    buffer_owner_.assign(std::cbegin(iter), std::cend(iter));
+    buffer_ = buffer_owner_;
+  }
+
+  /**
+   * Assigns a new non-owned buffer to the object.
+   *
+   * @param iter The iterator range to wrap in the serialization buffer.
+   */
+  template <class Iter>
+  inline void assign(NonOwnedMarker, const Iter& iter) {
+    // Clear vector and deallocate its buffer.
+    buffer_owner_.clear();
+    buffer_owner_.shrink_to_fit();
+    buffer_ = iter;
+  }
+
+  /**
+   * Assigns a new owned buffer to the object, and appends a null terminator.
+   */
+  template <class Iter>
+  inline void assign_null_terminated(const Iter& iter) {
+    // Clear vector and deallocate its buffer.
+    buffer_owner_.clear();
+    buffer_owner_.shrink_to_fit();
+    // Reserve enough space for the data and the null terminator.
+    buffer_owner_.reserve(
+        std::distance(std::cbegin(iter), std::cend(iter)) + 1);
+    buffer_owner_.assign(std::cbegin(iter), std::cend(iter));
+    buffer_owner_.push_back('\0');
+    buffer_ = buffer_owner_;
+  }
+
+  /**
+   * Gets the number of bytes in the buffer.
+   */
+  inline size_t size() const {
+    return buffer_.size();
+  }
+
+  /**
+   * Returns a mutable span to the buffer's whole data.
+   *
+   * This function must be called only on owned buffers, otherwise it will
+   * throw.
+   *
+   * The span returned by this function must not be used after one of the assign
+   * methods is called.
+   *
+   * @return A mutable span to the buffer's whole data.
+   */
+  inline span<char> owned_mutable_span() & {
+    if (!is_owned())
+      throw BufferStatusException(
+          "Cannot get a mutable span of a non-owned buffer.");
```

Comment:
I could just remove that or replace it with assert, which would always return an empty span for non-owned buffers.